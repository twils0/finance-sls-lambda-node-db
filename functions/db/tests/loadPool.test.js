const { getSecret } = require('../../aws/getSecret');
const { Pool } = require('pg');

const { loadPool } = require('../loadPool');

global.console.log = jest.fn();
global.process.exit = jest.fn();
jest.mock('../../aws/getSecret', () => ({
  getSecret: jest.fn(),
}));
jest.mock('pg', () => ({
  Pool: jest.fn(),
}));
const pool = {
  on: jest.fn(),
};

const getSecretValue = {
  key1: 'value1',
  key2: 'value2',
};
const poolConfig = {
  ...getSecretValue,
  min: 0,
  max: 1,
  idleTimeoutMillis: 1,
  connectionTimeoutMillis: 1000,
};

describe('functions', () => {
  describe('db', () => {
    describe('loadPool', () => {
      afterEach(() => {
        getSecret.mockReset();
        Pool.mockReset();
        pool.on.mockReset();
        global.process.exit.mockReset();
        global.console.log.mockReset();
      });

      it('fails and returns unexpected error, when getSecret throws error', async () => {
        const error = {
          code: 'testCode',
          message: 'testMessage',
        };
        getSecret.mockReturnValue(Promise.reject(error));

        try {
          await loadPool();
        } catch (errorCatch) {
          expect(errorCatch).toEqual(error);
        }
      });

      it('fails and returns unexpected error, when getSecret does not return a value', async () => {
        getSecret.mockReturnValue(Promise.resolve());

        try {
          await loadPool();
        } catch (errorCatch) {
          expect(errorCatch).toMatchSnapshot();
        }
      });

      it('fails and returns unexpected error, when pool throws error', async () => {
        const error = {
          code: 'testCode',
          message: 'testMessage',
        };
        getSecret.mockReturnValue(Promise.resolve({ value: getSecretValue }));
        pool.on.mockImplementation((type, callback) => callback(error));
        Pool.mockImplementation((config) => {
          expect(config).toEqual(poolConfig);
          return pool;
        });

        try {
          await loadPool();
        } catch (errorCatch) {
          expect(errorCatch).toMatchSnapshot();
        }

        expect(global.pool.on).toBeCalled();
        expect(global.process.exit).toBeCalledWith(-1);
        expect(global.console.log).toMatchSnapshot();
      });

      it('correctly loads db', async () => {
        getSecret.mockReturnValue(Promise.resolve({ value: getSecretValue }));
        Pool.mockImplementation((config) => {
          expect(config).toEqual(poolConfig);
          return pool;
        });

        const result = await loadPool();

        expect(result).toEqual(null);
        expect(global.console.log).toMatchSnapshot();
      });
    });
  });
});
