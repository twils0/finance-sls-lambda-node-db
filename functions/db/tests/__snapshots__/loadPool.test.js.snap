// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`functions db loadPool correctly loads db 1`] = `[MockFunction]`;

exports[`functions db loadPool fails and returns unexpected error, when getSecret does not return a value 1`] = `"unexpected error - unable to access db credentials from getSecret"`;

exports[`functions db loadPool fails and returns unexpected error, when pool throws error 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "unexpected error - most likely idle client; function will end
",
      Object {
        "code": "testCode",
        "message": "testMessage",
      },
    ],
    Array [
      "DB Loaded
",
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": undefined,
    },
    Object {
      "isThrow": false,
      "value": undefined,
    },
  ],
}
`;
